### PR TITLE
[RF] Support ShapeSys in JSON tool

### PR DESCRIPTION
**Pull request by gartrog**
Refactor ParamHistFunc treatment to support both MC stat (BB-lite)
and user-defined ShapeSys

Validated on ATLAS VHbb workspace

### CODE REVIEW

**Comment by gartrog** [comment]
Tagging @cburgard for review

**Comment by cburgard** [comment]
Looks good to me. Only one small suggestion: The `createPHF` function (which was originally written to only deal with mcstat) has been generalized by you to also work for ShapeSys, which is great. I think it would be good if the parameter names would reflect that (i.e., rename `statErrorType` to `constraintType` or similar). 

In addition to that, I'm sure @guitargeek will request you run `clang-format -i --style=file roofit/hs3/src/*.cxx`.

Thanks a lot for going the extra mile of adding this!

**Comment by gartrog** [comment]
Done.

**Comment by guitargeek** [comment]
Hi @gartrog, thanks for the PR! I see that it collides with my memory leak PR: https://github.com/root-project/root/pull/9690

Is it okay if I merge the memory management PR first, and then we come back to this?

**Comment by gartrog** [comment]
Hi @guitargeek yes please merge the other one first. It should be easy to adapt this one afterwards.

**Comment by guitargeek** [comment]
Ok! The conflicting PR has now been merged.

**Comment by gartrog** [comment]
It should now be ok, @guitargeek

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/hs3/src/RooJSONFactoryWSTool.cxx [hunk 1]
```diff
@@ -495,8 +495,8 @@ std::vector<std::vector<int>> RooJSONFactoryWSTool::generateBinIndices(const Roo
    std::vector<std::vector<int>> combinations;
    std::vector<int> vars_numbins;
    vars_numbins.reserve(vars.size());
-   for (auto &absv : vars) {
-      vars_numbins.push_back(((RooRealVar *)absv)->numBins());
+   for (const auto &absv : static_range_cast<RooRealVar *>(vars)) {
```
Comment (guitargeek) [inline roofit/hs3/src/RooJSONFactoryWSTool.cxx:498]:
```suggestion
   for (const auto *absv : static_range_cast<RooRealVar *>(vars)) {
```
Some compilers in the CI complain about references to a temporary pointer here. If you use the `range_cast` with a pointer type, you also need to copy that pointer with `const auto*`.

**Review by guitargeek — APPROVED** [review]
LGTM, thank you very much for implementing this!

The compiler warnings seen in the CI tests will be fixed by https://github.com/root-project/root/pull/9734.

**Pull request merged**